### PR TITLE
Allow manual reload of nix environment.

### PR DESCRIPTION
**Pull request by pelme**
Adding nix_direnv_manual_reload to .envrc will avoid the automatic update of the nix environment on changes and allow the user to issue a command to update the nix environment explicitly.

Sometimes it takes a long time to rebuild the nix environment and it happens in times. Switching branches in or rebasing can be painful since nix will re-evaluate between nix file changes.

This blog post explains the problems with blocking the shell: https://dev.to/allenap/some-direnv-best-practices-actually-just-one-4864

Example usage:

```
$ cat .envrc
source nix-direnv/direnvrc

nix_direnv_manual_reload
use_flake

$ touch flake.nix
direnv: loading ~/tst/.envrc
direnv: nix-direnv: cache is out of date. use "nix-direnv-reload" to reload.
direnv: export +AR +AS <... snip ...>

 $ nix-direnv-reload
direnv: loading ~/tst/.envrc
direnv: nix-direnv: renewed cache
```

What do you think about this?

### CODE REVIEW

**Comment by pelme** [comment]
As for improved UX, what about creating a script at `.direnv/bin/nix-direnv-refresh` and using `PATH_add .direnv/bin`? The script would contain:
```
#!/usr/bin/env bash
nix_direnv_refresh=t direnv exec $direnvpath true
```

To refresh the cache, the command would then be:
```
$ nix-direnv-refresh
```

**Comment by pelme** [comment]
> As for improved UX, what about creating a script at `.direnv/bin/nix-direnv-refresh` and using `PATH_add .direnv/bin`? The script would contain:
> 
> ```
> #!/usr/bin/env bash
> nix_direnv_refresh=t direnv exec $direnvpath true
> ```
> 
> To refresh the cache, the command would then be:
> 
> ```
> $ nix-direnv-refresh
> ```

I have implemented this idea and updated the PR.

**Comment by phaer** [comment]
> I have implemented this idea and updated the PR.

I like the idea, but I think `nix-direnv-refresh` could be a simple shell function instead of a script on disk, as it's only ever needed in a shell session where your init code already ran?

**Comment by Mic92** [comment]
> > I have implemented this idea and updated the PR.
> 
> I like the idea, but I think `nix-direnv-refresh` could be a simple shell function instead of a script on disk, as it's only ever needed in a shell session where your init code already ran?

You cannot have shell functions because direnv only exports environment variables back into the shell. This is because direnv wants to support non-posix shells like fish.

**Review by bbenne10 — COMMENTED** [review]
With the changes I've asked for, I'd be happy to merge this.

**Comment by bbenne10** [comment]
Note that you'll also need to address the shellcheck failures before merge.

File: direnvrc
Comment (Mic92) [reply direnvrc:167]: I would also make scope it, since it's a global variable:

```suggestion
nix_direnv_auto_reload=t
```

File: direnvrc
Comment (Mic92) [inline direnvrc:208]:
```suggestion
    if [[ ! -x "${layout_dir}/bin/nix-direnv-reload" ]]; then chmod +x "${layout_dir}/bin/nix-direnv-reload"; fi
```

File: direnvrc
Comment (Mic92) [reply direnvrc:204]: I'd say:

```
echo > "${layout_dir}/bin/nix-direnv-reload" '#!/usr/bin/env bash
dir="$(realpath $(dirname ${BASH_SOURCE[0]})/../..)"
nix_direnv_reload=t direnv exec "$dir" true
direnv reload
'
```

**Comment by pelme** [comment]
Thanks a lot for your feedback and I am glad you find it useful. Bash scripting is not my primary cup of tea and I appreciate the improvements you have suggested. Will go through all comments and update the PR! Will also make sure use_nix works too, I did only focus on use_flake to start of with.

**Comment by pelme** [comment]
I have updated the PR with these changes:

- Always create the nix-direnv-reload script
- Use 0/1 for booleans
- Avoid crashing when profile_rc does not exist the first time
- Use better internal variable name for manual mode
- Make use_nix work the same as use_flakes
- Avoid unneccessary forks where possible
- Fix shellcheck failures
- Update mtime on profile rc files to avoid confusion about if the files are up to date
- Add instructions to README.md

**Review by bbenne10 — APPROVED** [review]
Lgtm. Since @Mic92 also requested changes, I'll wait till he approves as well to merge.

**Comment by Mic92** [comment]
@mergify queue

**Comment by Mic92** [comment]
@mergifyio refresh

**Comment by Mic92** [comment]
@Mergifyio rebase

**Comment by Mic92** [comment]
@mergify queue

**Pull request merged**
